File: 145/BinaryTreePostOrderTraversal.py
# https://leetcode.com/problems/binary-tree-postorder-traversal/description/?envType=daily-question&envId=2024-08-25

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def postorderTraversal(self, root: Optional[TreeNode]) -> List[int]:

        def depthFirstSearch_forTrees(root, resultantPostOrderList):
            if root is None:
                return
            depthFirstSearch_forTrees(root.left, resultantPostOrderList)
            depthFirstSearch_forTrees(root.right, resultantPostOrderList)
            if root is not None:
                resultantPostOrderList.append(root.val)
            return

        resultantPostOrderList=[]
        depthFirstSearch_forTrees(root, resultantPostOrderList)
        return resultantPostOrderList

File: 145/BinaryTreePostOrderTraversal.cpp
/*
https://leetcode.com/problems/binary-tree-postorder-traversal/description/?envType=daily-question&envId=2024-08-25
*/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:

    void dfs(TreeNode *root, vector<int>&res){
        if(root==nullptr) return;
        dfs(root->left, res);
        dfs(root->right, res);
        if(root!=nullptr){
            res.push_back(root->val);
        }
        return;
    }

    vector<int> postorderTraversal(TreeNode* root) {
        vector<int>res;
        dfs(root, res);
        return res;
    }
};